Sources/Commands/PinsStore.swift

```diff
@@ -0,0 +1,188 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Basic
+import SourceControl
+import Utility
+
+public enum PinOperationError: Swift.Error {
+    case alreadyPinned
+    case notPinned
+}
+
+public struct PinsStore {
+    public struct Pin {
+        /// The repository specifer of the pinned dependency.
+        public let repository: RepositorySpecifier
+
+        /// The pinned version.
+        public let version: Version
+
+        /// The reason text for pinning this dependency.
+        public let reason: String?
+
+        init(repository: RepositorySpecifier, version: Version, reason: String? = nil) {
+            self.repository = repository
+            self.version = version
+            self.reason = reason
+        }
+    }
+
+    /// The path to the pins file.
+    fileprivate let pinsFile: AbsolutePath
+
+    /// The filesystem to manage the pin file on.
+    fileprivate var fileSystem: FileSystem
+
+    /// The pins map.
+    fileprivate(set) var pinsMap: [RepositorySpecifier: Pin]
```

Comment:
It is unique, but it can change (https -> git, or case changes, etc.) while pin probably should remain independent of that.

So I like idea of using the package name as the more normalized form.